fili-core/src/main/java/com/yahoo/bard/webservice/data/config/table/PhysicalTableDefinition.java

```diff
@@ -1,93 +1,144 @@
-
 import java.util.Collections;
+import java.util.LinkedHashSet;
 import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 /**
- * Holds the fields needed to define a Physical Table.
+ * Common configurations necessary to build a physical table.
  */
-public class PhysicalTableDefinition {
+public abstract class PhysicalTableDefinition {
+
     private final TableName name;
-    private final ZonedTimeGrain grain;
-    private final ImmutableSet<? extends DimensionConfig> dimensions;
+    private final ZonedTimeGrain timeGrain;
+    private final Set<FieldName> metricNames;
+    private final Set<? extends DimensionConfig> dimensionConfigs;
     private final Map<String, String> logicalToPhysicalNames;
 
     /**
-     * Define a physical table using a zoned time grain.
+     * Constructor for sub-class to call.
      *
-     * @param name  The table name
-     * @param grain  The zoned time grain
-     * @param dimensions  The dimension configurations
+     * @param name  Table name of the physical table
+     * @param timeGrain  Zoned time grain of the table
+     * @param metricNames  The Set of metric names on the table
+     * @param dimensionConfigs  Set of dimensions on the table as dimension configs
      */
-    public PhysicalTableDefinition(
+    protected PhysicalTableDefinition(
             TableName name,
-            ZonedTimeGrain grain,
-            Iterable<? extends DimensionConfig> dimensions
+            ZonedTimeGrain timeGrain,
+            Set<FieldName> metricNames,
+            Set<? extends DimensionConfig> dimensionConfigs
     ) {
         this.name = name;
-        this.grain = grain;
-        this.dimensions = ImmutableSet.copyOf(dimensions);
-        this.logicalToPhysicalNames = Collections.unmodifiableMap(
-                this.dimensions.stream()
-                        .collect(
-                                Collectors.toMap(
-                                        DimensionConfig::getApiName,
-                                        dim -> {
-                                            String physicalName = dim.getPhysicalName();
-                                            if (physicalName == null) {
-                                                throw new RuntimeException("No physical name found for dimension: "
-                                                        + dim.getApiName());
-                                            }
-                                            return physicalName;
-                                        }
-                                )
-                        )
-        );
+        this.timeGrain = timeGrain;
+        this.metricNames = ImmutableSet.copyOf(metricNames);
+        this.dimensionConfigs = ImmutableSet.copyOf(dimensionConfigs);
+        this.logicalToPhysicalNames = Collections.unmodifiableMap(buildLogicalToPhysicalNames(dimensionConfigs));
```

Comment:
If we really want this class to be primarily a holder for data needed to build a physical table, we need a constructor that takes this `logicalToPhysicalNames` mapping in as a parameter, rather than just assuming we can build it from the dimension configs.

The reason for this stems from the reason for the mapping in the first place: Dimensions do not have the same physical name on all physical tables. Because of that, you'll either need to have a complete set of DimensionConfig objects for every table, or you need to be able to pass the map into the PTDefinition directly.

(Note that this concern likely ripples down to the constructors of the subclasses as well.